File: experiments/problem1_nonperiodicity/analyze_period_two_return_penalty_locality.cpp
#include <array>
#include <cstdint>
#include <iostream>
#include <map>
#include <stdexcept>
#include <string>
#include <unordered_map>
#include <utility>
#include <vector>

namespace {

constexpr std::array<char, 3> LETTERS{'t', 'p', 'u'};
constexpr std::array<char, 2> PHASES{'p', 'u'};
constexpr std::array<int, 4> GAPS{2, 3, 4, 5};
constexpr int MAXIMUM_FINAL_DEPTH = 12;

constexpr std::array<int, 23> KAPPA_P{
    0, 1, 3, 7, 8, 8, 12, 13, 17, 17, 17, 21, 28, 30, 33, 34, 36, 40,
    40, 42, 47, 49, 51,
};
constexpr std::array<int, 23> KAPPA_U{
    0, 2, 2, 2, 7, 12, 14, 14, 14, 18, 19, 26, 27, 30, 30, 30, 30, 40,
    42, 42, 49, 52, 52,
};

uint32_t width_mask(int width) {
    if (width < 0 || width > 30) {
        throw std::invalid_argument("controlled width outside 0..30");
    }
    return width == 0 ? 0 : ((uint32_t{1} << width) - 1);
}

uint32_t forward_generator(char letter, uint32_t state, int width) {
    const uint32_t stepped =
        (state ^ ((state << 1) | (state << 2))) & width_mask(width);
    if (letter == 't') return stepped;
    if (letter == 'u') return stepped ^ 1;
    if (letter == 'p') {
        return stepped ^ 1 ^ ((state & 1) == 0 ? 2 : 0);
    }
    throw std::invalid_argument("unknown generator");
}

uint32_t inverse_t_mod(uint32_t output, int width) {
    output &= width_mask(width);
    uint32_t state = 0;
    for (int position = 0; position < width; ++position) {
        uint32_t previous = 0;
        if (position >= 1) previous |= (state >> (position - 1)) & 1;
        if (position >= 2) previous |= (state >> (position - 2)) & 1;
        const uint32_t bit = ((output >> position) & 1) ^ previous;
        state |= bit << position;
    }
    return state;
}

uint32_t inverse_generator_mod(char letter, uint32_t output, int width) {
    output &= width_mask(width);
    if (letter == 't') return inverse_t_mod(output, width);
    if (letter == 'u') return inverse_t_mod(output ^ 1, width);
    if (letter == 'p') {
        const uint32_t recovered_low_bit = (output & 1) ^ 1;
        const uint32_t adjusted =
            output ^ 1 ^ (recovered_low_bit == 0 ? 2 : 0);
        return inverse_t_mod(adjusted, width);
    }
    throw std::invalid_argument("unknown inverse generator");
}

uint32_t backward_zero_branch(char branch, uint32_t successor, int width) {
    const int inner_width = width - 2;
    uint32_t state = inverse_generator_mod(branch, successor, inner_width);
    state = inverse_generator_mod('p', state, inner_width);
    return ((state << 2) | 3) & width_mask(width);
}

uint32_t survivor_for_prefix(const std::string& prefix) {
    uint32_t state = 0;
    int precision = 0;
    for (auto iterator = prefix.rbegin(); iterator != prefix.rend(); ++iterator) {
        precision += 2;
        state = backward_zero_branch(*iterator, state, precision);
    }
    return state;
}

bool locally_admissible(const std::string& word) {
    return word.find("uu") == std::string::npos
        && word.find("ttttt") == std::string::npos
        && word.find("ututtu") == std::string::npos;
}

void generate_prefixes(
    int length,
    std::string& prefix,
    std::vector<std::string>& output
) {
    if (static_cast<int>(prefix.size()) == length) {
        if (!prefix.empty() && prefix.back() == 'u') output.push_back(prefix);
        return;
    }
    for (char letter : {'t', 'u'}) {
        prefix.push_back(letter);
        if (locally_admissible(prefix)) {
            generate_prefixes(length, prefix, output);
        }
        prefix.pop_back();
    }
}

std::vector<uint8_t> phase_distances(int depth, char phase) {
    const int width = 2 * depth;
    const uint32_t state_count = uint32_t{1} << width;
    const uint32_t start = forward_generator(phase, 0, width);
    std::vector<uint8_t> distance(state_count, 255);
    std::vector<uint32_t> queue;
    queue.reserve(state_count);
    distance[start] = 0;
    queue.push_back(start);
    for (std::size_t head = 0; head < queue.size(); ++head) {
        const uint32_t state = queue[head];
        const uint8_t next_distance = distance[state] + 1;
        for (char letter : LETTERS) {
            const uint32_t image = forward_generator(letter, state, width);
            if (distance[image] == 255) {
                distance[image] = next_distance;
                queue.push_back(image);
            }
        }
    }
    return distance;
}

struct Summary {
    int candidates = 0;
    int zero = 0;
    int positive = 0;
    int minimum = 255;
    int maximum = -1;
    std::map<int, int> histogram;
    std::string first_zero_prefix;
    std::string first_positive_prefix;
    int first_zero_depth = -1;
    int first_positive_depth = -1;
};

void print_string(const std::string& value) {
    std::cout << '"' << value << '"';
}

void print_summary(const Summary& summary) {
    std::cout << "{\"candidates\":" << summary.candidates
              << ",\"zero_penalty_count\":" << summary.zero
              << ",\"positive_penalty_count\":" << summary.positive
              << ",\"minimum_penalty\":" << summary.minimum
              << ",\"maximum_penalty\":" << summary.maximum
              << ",\"histogram\":{";
    bool first = true;
    for (const auto& [penalty, count] : summary.histogram) {
        if (!first) std::cout << ',';
        first = false;
        std::cout << '"' << penalty << "\":" << count;
    }
    std::cout << "},\"first_zero_example\":";
    if (summary.first_zero_depth < 0) {
        std::cout << "null";
    } else {
        std::cout << "{\"base_depth\":" << summary.first_zero_depth
                  << ",\"prefix\":";
        print_string(summary.first_zero_prefix);
        std::cout << '}';
    }
    std::cout << ",\"first_positive_example\":";
    if (summary.first_positive_depth < 0) {
        std::cout << "null";
    } else {
        std::cout << "{\"base_depth\":" << summary.first_positive_depth
                  << ",\"prefix\":";
        print_string(summary.first_positive_prefix);
        std::cout << '}';
    }
    std::cout << '}';
}

}  // namespace

int main() {
    std::unordered_map<int, std::vector<uint8_t>> p_distance;
    std::unordered_map<int, std::vector<uint8_t>> u_distance;
    for (int depth = 1; depth <= MAXIMUM_FINAL_DEPTH; ++depth) {
        p_distance.emplace(depth, phase_distances(depth, 'p'));
        u_distance.emplace(depth, phase_distances(depth, 'u'));
    }

    std::cout << "{\"maximum_final_depth\":" << MAXIMUM_FINAL_DEPTH
              << ",\"known_forbidden_factors\":[\"uu\",\"ttttt\",\"ututtu\"]"
              << ",\"return_gap_rows\":{";
    bool first_gap = true;
    for (int gap : GAPS) {
        if (!first_gap) std::cout << ',';
        first_gap = false;
        std::cout << '"' << gap << "\":{";
        bool first_phase = true;
        const std::string extension(gap - 1, 't');
        for (char phase : PHASES) {
            if (!first_phase) std::cout << ',';
            first_phase = false;
            Summary summary;
            for (int base_depth = 1;
                 base_depth + gap <= MAXIMUM_FINAL_DEPTH;
                 ++base_depth) {
                std::vector<std::string> prefixes;
                std::string prefix;
                generate_prefixes(base_depth, prefix, prefixes);
                for (const std::string& candidate : prefixes) {
                    const std::string extended = candidate + extension + 'u';
                    if (!locally_admissible(extended)) continue;
                    const uint32_t current_target = survivor_for_prefix(candidate);
                    const uint32_t following_target = survivor_for_prefix(extended);
                    const int current = 1 + (
                        phase == 'p'
                            ? p_distance.at(base_depth)[current_target]
                            : u_distance.at(base_depth)[current_target]
                    );
                    const int final_depth = base_depth + gap;
                    const int following = 1 + (
                        phase == 'p'
                            ? p_distance.at(final_depth)[following_target]
                            : u_distance.at(final_depth)[following_target]
                    );
                    const int penalty = following - current;
                    if (penalty < 0) {
                        throw std::runtime_error("projection monotonicity failed");
                    }
                    ++summary.candidates;
                    ++summary.histogram[penalty];
                    summary.minimum = std::min(summary.minimum, penalty);
                    summary.maximum = std::max(summary.maximum, penalty);
                    if (penalty == 0) {
                        ++summary.zero;
                        if (summary.first_zero_depth < 0) {
                            summary.first_zero_depth = base_depth;
                            summary.first_zero_prefix = candidate;
                        }
                    } else {
                        ++summary.positive;
                        if (summary.first_positive_depth < 0) {
                            summary.first_positive_depth = base_depth;
                            summary.first_positive_prefix = candidate;
                        }
                    }
                }
            }
            std::cout << '"' << phase << "\":";
            print_summary(summary);
        }
        std::cout << '}';
    }
    std::cout << "},\"actual_exact_returns_through_depth_21\":[";
    constexpr std::array<int, 7> POSITIONS{0, 4, 6, 11, 13, 18, 20};
    for (std::size_t index = 0; index + 1 < POSITIONS.size(); ++index) {
        if (index) std::cout << ',';
        const int base_depth = POSITIONS[index] + 1;
        const int final_depth = POSITIONS[index + 1] + 1;
        std::cout << "{\"return_index\":" << index
                  << ",\"gap\":" << POSITIONS[index + 1] - POSITIONS[index]
                  << ",\"base_depth\":" << base_depth
                  << ",\"final_depth\":" << final_depth
                  << ",\"p_penalty\":" << KAPPA_P[final_depth] - KAPPA_P[base_depth]
                  << ",\"u_penalty\":" << KAPPA_U[final_depth] - KAPPA_U[base_depth]
                  << '}';
    }
    std::cout << "]}\n";
    return 0;
}

File: experiments/problem1_nonperiodicity/analyze_period_two_return_penalty_locality.py
#!/usr/bin/env python3
"""Classify local return-gap penalties for period-two phase witnesses.

For a finite branch prefix ``w`` ending in ``u`` and a return gap ``r``, append
``t**(r-1) u``.  The exact phase penalty is the increase in positive-generator
distance from the phase start to the corresponding schedule-survivor cylinder.

The finite campaign exhausts prefixes obeying the three proved fringe-language
exclusions.  It is regression evidence for a sharper all-depth lemma; it does
not itself prove that every non-two return has positive penalty.
"""
from __future__ import annotations

import argparse
import hashlib
import json
from array import array
from collections import Counter
from functools import lru_cache
from typing import Any, Iterator

LETTERS = ("t", "p", "u")
PHASES = ("p", "u")
RETURN_GAPS = (2, 3, 4, 5)
FORBIDDEN = ("uu", "ttttt", "ututtu")
DEFAULT_MAXIMUM_FINAL_DEPTH = 9
ABSOLUTE_MAXIMUM_FINAL_DEPTH = 10

KAPPA_P = (
    0, 1, 3, 7, 8, 8, 12, 13, 17, 17, 17, 21, 28, 30, 33, 34, 36, 40,
    40, 42, 47, 49, 51,
)
KAPPA_U = (
    0, 2, 2, 2, 7, 12, 14, 14, 14, 18, 19, 26, 27, 30, 30, 30, 30, 40,
    42, 42, 49, 52, 52,
)


class ReturnPenaltyLimitError(RuntimeError):
    """Raised before a controlled exhaustive campaign exceeds its limit."""


def width_mask(width: int) -> int:
    if width < 0:
        raise ValueError("width must be nonnegative")
    return (1 << width) - 1 if width else 0


def forward_generator(name: str, state: int, width: int) -> int:
    stepped = (state ^ ((state << 1) | (state << 2))) & width_mask(width)
    if name == "t":
        return stepped
    if name == "u":
        return stepped ^ 1
    if name == "p":
        return stepped ^ 1 ^ (2 if state & 1 == 0 else 0)
    raise ValueError("unknown generator")


def inverse_t_mod(output: int, width: int) -> int:
    output &= width_mask(width)
    state = 0
    for position in range(width):
        previous = 0
        if position >= 1:
            previous |= (state >> (position - 1)) & 1
        if position >= 2:
            previous |= (state >> (position - 2)) & 1
        bit = ((output >> position) & 1) ^ previous
        state |= bit << position
    return state


def inverse_generator_mod(name: str, output: int, width: int) -> int:
    output &= width_mask(width)
    if name == "t":
        return inverse_t_mod(output, width)
    if name == "u":
        return inverse_t_mod(output ^ 1, width)
    if name == "p":
        recovered_low_bit = (output & 1) ^ 1
        adjusted = output ^ 1 ^ (2 if recovered_low_bit == 0 else 0)
        return inverse_t_mod(adjusted, width)
    raise ValueError("unknown inverse generator")


def backward_zero_branch(branch: str, successor: int, width: int) -> int:
    inner_width = width - 2
    state = inverse_generator_mod(branch, successor, inner_width)
    state = inverse_generator_mod("p", state, inner_width)
    return ((state << 2) | 3) & width_mask(width)


def survivor_for_prefix(prefix: str) -> int:
    state = 0
    precision = 0
    for branch in reversed(prefix):
        precision += 2
        state = backward_zero_branch(branch, state, precision)
    return state


def is_locally_admissible(word: str) -> bool:
    return not any(factor in word for factor in FORBIDDEN)


def valid_prefixes_ending_u(length: int) -> Iterator[str]:
    if length <= 0:
        return

    def visit(prefix: str) -> Iterator[str]:
        if len(prefix) == length:
            if prefix.endswith("u"):
                yield prefix
            return
        for letter in ("t", "u"):
            candidate = prefix + letter
            if is_locally_admissible(candidate):
                yield from visit(candidate)

    yield from visit("")


@lru_cache(maxsize=None)
def phase_distances(depth: int, phase: str) -> bytes:
    if depth <= 0:
        raise ValueError("depth must be positive")
    if phase not in PHASES:
        raise ValueError("phase must be p or u")
    width = 2 * depth
    state_count = 1 << width
    start = forward_generator(phase, 0, width)
    distances = bytearray(b"\xff") * state_count
    queue = array("I", [start])
    distances[start] = 0
    head = 0
    while head < len(queue):
        state = queue[head]
        head += 1
        next_distance = distances[state] + 1
        for letter in LETTERS:
            image = forward_generator(letter, state, width)
            if distances[image] == 255:
                distances[image] = next_distance
                queue.append(image)
    return bytes(distances)


def phase_complexity(prefix: str, phase: str) -> int:
    target = survivor_for_prefix(prefix)
    return 1 + phase_distances(len(prefix), phase)[target]


def classify_gap(gap: int, phase: str, maximum_final_depth: int) -> dict[str, Any]:
    extension = "t" * (gap - 1) + "u"
    histogram: Counter[int] = Counter()
    zero_examples: list[dict[str, Any]] = []
    positive_example: dict[str, Any] | None = None
    candidates = 0

    for base_depth in range(1, maximum_final_depth - gap + 1):
        for prefix in valid_prefixes_ending_u(base_depth):
            extended = prefix + extension
            if not is_locally_admissible(extended):
                continue
            current = phase_complexity(prefix, phase)
            following = phase_complexity(extended, phase)
            penalty = following - current
            if penalty < 0:
                raise AssertionError("projection monotonicity failed")
            candidates += 1
            histogram[penalty] += 1
            row = {
                "base_depth": base_depth,
                "prefix": prefix,
                "extended_prefix": extended,
                "current_complexity": current,
                "following_complexity": following,
                "penalty": penalty,
            }
            if penalty == 0 and len(zero_examples) < 4:
                zero_examples.append(row)
            if penalty > 0 and positive_example is None:
                positive_example = row

    return {
        "candidates": candidates,
        "zero_penalty_count": histogram[0],
        "positive_penalty_count": candidates - histogram[0],
        "minimum_penalty": min(histogram),
        "maximum_penalty": max(histogram),
        "histogram": {str(key): histogram[key] for key in sorted(histogram)},
        "zero_examples": zero_examples,
        "positive_example": positive_example,
    }


def actual_return_rows() -> list[dict[str, Any]]:
    positions = (0, 4, 6, 11, 13, 18, 20)
    rows = []
    for index, (left, right) in enumerate(zip(positions, positions[1:])):
        base_depth = left + 1
        final_depth = right + 1
        rows.append(
            {
                "return_index": index,
                "gap": right - left,
                "base_depth": base_depth,
                "final_depth": final_depth,
                "p_penalty": KAPPA_P[final_depth] - KAPPA_P[base_depth],
                "u_penalty": KAPPA_U[final_depth] - KAPPA_U[base_depth],
            }
        )
    return rows


def run_campaign(
    maximum_final_depth: int = DEFAULT_MAXIMUM_FINAL_DEPTH,
) -> dict[str, Any]:
    if not 6 <= maximum_final_depth <= ABSOLUTE_MAXIMUM_FINAL_DEPTH:
        raise ReturnPenaltyLimitError("maximum final depth outside controlled range")

    rows: dict[str, Any] = {}
    for gap in RETURN_GAPS:
        rows[str(gap)] = {
            phase: classify_gap(gap, phase, maximum_final_depth)
            for phase in PHASES
        }

    payload: dict[str, Any] = {
        "status": "finite-exhaustive",
        "maximum_final_depth": maximum_final_depth,
        "known_forbidden_factors": list(FORBIDDEN),
        "return_gap_rows": rows,
        "actual_exact_returns_through_depth_21": actual_return_rows(),
        "all_depth_reduction": {
            "conditional_lemma": (
                "if every locally admissible return of gap 3, 4, or 5 has positive "
                "phase penalty, then bounded phase complexity forces all sufficiently "
                "late return gaps to equal 2"
            ),
            "periodic_consequence": (
                "eventual gap 2 makes the branch schedule eventually repeat ut; the "
                "existing schedule-coding theorem excludes an ordinary finite survivor "
                "for every eventually periodic schedule"
            ),
        },
        "scope_warning": (
            "the bounded campaign does not prove all-depth positivity for gaps 3, 4, "
            "or 5, does not exclude period two, and does not solve Rule 30 center "
            "nonperiodicity"
        ),
    }
    encoded = json.dumps(payload, sort_keys=True, separators=(",", ":")).encode()
    payload["certificate_sha256"] = hashlib.sha256(encoded).hexdigest()
    return payload


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--maximum-final-depth",
        type=int,
        default=DEFAULT_MAXIMUM_FINAL_DEPTH,
    )
    args = parser.parse_args()
    print(json.dumps(run_campaign(args.maximum_final_depth), indent=2, sort_keys=True))


if __name__ == "__main__":
    main()
